rust/src/buffer.rs

```diff
@@ -15,278 +15,170 @@
 // specific language governing permissions and limitations
 // under the License.
 
-use bytes::Bytes;
-use libc;
 use std::mem;
-use std::slice;
+use std::sync::Arc;
 
-use super::datatypes::*;
-use super::memory::*;
+use memory;
 
-/// Buffer<T> is essentially just a Vec<T> for fixed-width primitive types and the start of the
-/// memory region is aligned at a 64-byte boundary
-pub struct Buffer<T>
-where
-    T: ArrowPrimitiveType,
-{
-    /// Contiguous memory region holding instances of primitive T
-    data: *const T,
-    /// Number of elements in the buffer
-    len: usize,
+/// Buffer is a contiguous memory region of fixed size and is aligned at a 64-byte
+/// boundary. Buffer is immutable.
+#[derive(PartialEq, Debug)]
+pub struct Buffer {
+    /// Reference-counted pointer to the internal byte buffer.
+    data: Arc<BufferData>,
+
+    /// The offset into the buffer.
+    offset: usize,
 }
 
-impl<T> Buffer<T>
-where
-    T: ArrowPrimitiveType,
-{
-    /// create a buffer from an existing region of memory (must already be byte-aligned)
-    pub unsafe fn from_raw_parts(data: *const T, len: usize) -> Self {
-        Buffer { data, len }
-    }
+#[derive(Debug)]
+struct BufferData {
+    /// The raw pointer into the buffer bytes
+    ptr: *const u8,
 
-    /// Get the number of elements in the buffer
-    pub fn len(&self) -> usize {
-        self.len
-    }
-
-    /// Get a pointer to the data contained by the buffer
-    pub fn data(&self) -> *const T {
-        self.data
-    }
+    /// The length of the buffer
+    len: usize,
+}
 
-    pub fn slice(&self, start: usize, end: usize) -> &[T] {
-        assert!(end <= self.len);
-        assert!(start <= end);
-        unsafe { slice::from_raw_parts(self.data.offset(start as isize), end - start) }
+impl PartialEq for BufferData {
+    fn eq(&self, other: &BufferData) -> bool {
+        unsafe { memory::memcmp(self.ptr, other.ptr, self.len as usize) == 0 }
```

Comment:
I think this needs to compare `self.len` to `other.len` first before calling `memcmp`?

Reply:
Good catch. Done.